MPT

miha-stopar opened the pull request:
This PR aims to provide the first version of Merkle Patricia Trie circuit. The witnesses are currently generated [here](https://github.com/miha-stopar/mpt). Lookups in the Keccak circuit are simulated using internal lookup table.

This branch replaces https://github.com/appliedzkp/zkevm-circuits/commits/merkle-patricia-trie (didn't rebase as it seems cleaner this way, will leave the old branch for a while before deleting it).

miha-stopar committed:
account drifted leaf RLC computation now takes into account new nonce_stored and balance_stored values

miha-stopar committed:
fix

miha-stopar committed:
account_leaf_key after placeholder branch fix

miha-stopar committed:
drifted leaf constraints when account deleted

miha-stopar committed:
PoisonedConstraint fix

miha-stopar committed:
not_first_level doubling removing from drifted leaf

miha-stopar committed:
missing branch init constraints

miha-stopar committed:
extension node missing constraints: range 256 lookups & RLP; two tests added

miha-stopar committed:
extension_node_key: constraints to avoid PoisonedConstraint removed

miha-stopar committed:
account delete selector

miha-stopar committed:
account delete constraint

miha-stopar committed:
non_existing_account selector

miha-stopar committed:
non-existing account proof row added

miha-stopar committed:
AccountNonExisting chip

miha-stopar committed:
NonExistingAccount test

miha-stopar committed:
non_existing_account proof: constraint for the two addresses being different

miha-stopar committed:
account leaf key in first level fixed

miha-stopar committed:
NonExistingAccount tests regenerated

miha-stopar committed:
non_existing_account proof constraints (missing subcase: when only one account in trie)

miha-stopar committed:
tests regenerated

miha-stopar committed:
account_leaf_key fix; test for non existing account in trie first level

miha-stopar committed:
PoisonedConstraint fixed

miha-stopar committed:
test for non_existing_account after first level; bug fixed

miha-stopar committed:
tests with account in/after the first level

miha-stopar committed:
ProofVariables now uses default

miha-stopar committed:
account in first level fixed

miha-stopar committed:
account in first level / storage in first storage level test and fixing

miha-stopar committed:
storage leaf in first level test & fix

miha-stopar committed:
account after placeholder branch test & fix

miha-stopar committed:
Account branch placeholder in first level test added; tests regenerated

miha-stopar committed:
account with placeholder branch in first level test and fix

miha-stopar committed:
account address rlc fixed not to go in negative rows when account leaf after placeholder in first level

miha-stopar committed:
cleaning non existing account proof

miha-stopar committed:
non_existing_account test for the case when there is only one account in state trie

miha-stopar committed:
extension node in account proof first level test & fix

miha-stopar committed:
Extension node with three nibbles tests and fix

miha-stopar committed:
fix in account_non_existing proof

miha-stopar committed:
generalizing for non-hashed nodes

miha-stopar committed:
branch RLP length and RLC for branches that contain non-hashed nodes

miha-stopar committed:
branch_rlc generalized for non-hashed nodes

miha-stopar committed:
initial work on extension_node generalization for non-hashed branches

miha-stopar committed:
non-hashed nodes; descriptions for constants in param.rs

miha-stopar committed:
storage leaf in last level: leaf RLC

miha-stopar committed:
addressing Adria comments

miha-stopar committed:
storage leaf in last level missing constraints added

miha-stopar committed:
started on non-hashed leaf in parent

miha-stopar committed:
storage leaf last level RLP constraints; leaf hash in parent for non-hashed leaves

miha-stopar committed:
non-hashed node move for one position to the right in branch children to be compatible with hash RLC computation

miha-stopar committed:
checking there are 0s after non-hashed RLP ends

miha-stopar committed:
non-hashed branch in extension node moved for one position to the right

miha-stopar committed:
extension node non-hashed branch fix

miha-stopar committed:
MainCols struct introduced instead of rlp1, rlp2, advices

miha-stopar committed:
missing selector in for drifted leaf when leaf in first storage level

miha-stopar committed:
removed is_codehash_mod as it is not possible to directly change the code

miha-stopar committed:
Started adding documentation to account_leaf_nonce_balance

miha-stopar committed:
ProofTypeCols struct introduced

miha-stopar committed:
non-hashed branch in extension node

miha-stopar committed:
documenting account_leaf_nonce_balance

miha-stopar committed:
AccountLeafCols

miha-stopar committed:
AccountLeaf struct

miha-stopar committed:
extension node RLP for non-hashed branch

miha-stopar committed:
documenting account_leaf_nonce_balance

miha-stopar committed:
StorageLeafCols struct

miha-stopar committed:
StorageLeaf struct

miha-stopar committed:
extension_node non-hashed branch RLP constraints; non-hashed node lookup fix in branch_rlc

miha-stopar committed:
documenting nonce balance chip

miha-stopar committed:
minor update to account_leaf_nonce_balance

miha-stopar committed:
BranchCols struct

miha-stopar committed:
added TODO in extension_node_key

miha-stopar committed:
account leaf nonce balance chip removed (only config)

miha-stopar committed:
AccountLeafNonceBalance assign function

miha-stopar committed:
Branch struct

miha-stopar committed:
assign_row takes branch struct

miha-stopar committed:
non-hashed nodes tests changed

miha-stopar committed:
AccountLeafStorageCodehash chip -> config; some docs

miha-stopar committed:
account leaf storage codehash docs

miha-stopar committed:
AccumulatorCols and AccumulatorPair structs

miha-stopar committed:
storage leaf with one nibble (one level above last level)

miha-stopar committed:
notes about non-hashed nodes in first level

miha-stopar committed:
assign function for account_leaf_storage_codehash

miha-stopar committed:
started documenting account leaf key

miha-stopar committed:
DenoteCols struct

miha-stopar committed:
account key leaf removed redundancy; added comments

miha-stopar committed:
key_rlc/key_rlc_mult moved to AccumulatorCols

miha-stopar committed:
non-hashed branch in parent check

miha-stopar committed:
Account leaf key constraints reintroduced (were not redundant); comments added

miha-stopar committed:
removed unneccessary arguments in some configs

miha-stopar committed:
fix

miha-stopar committed:
fixed extension node assignment; extension node key constraints for short version simplified

miha-stopar committed:
account_key_leaf comments

miha-stopar committed:
account key leaf assign function

miha-stopar committed:
node_mult_diff_s/node_mult_diff_c now part of AccumulatorCols struct

miha-stopar committed:
mult_diff moved into AccumulatorCols struct

miha-stopar committed:
Non-hashed extension node in parent branch constraint

miha-stopar committed:
s_mod_node_rlc/c_mod_node_rlc moved to AccumulatorCols

miha-stopar committed:
account leaf key in added branch documentation

miha-stopar committed:
account leaf key in added branch assign function

miha-stopar committed:
Account leaf comments

miha-stopar committed:
MptWitnessRow & MptWitnessRowType introduced as proposed by Adria

miha-stopar committed:
Documenting account_non_existing

miha-stopar committed:
branch init assignment simplification as proposed by Adria

miha-stopar committed:
branch init assignment refactored

miha-stopar committed:
number of nibbles used assignment

miha-stopar committed:
key_len_lookup for account_non_existing

miha-stopar committed:
Documenting account_non_existing

miha-stopar committed:
account_non_existing assign function

miha-stopar committed:
account leaf structs in separate file

miha-stopar committed:
storage leaf structs in a separate file

miha-stopar committed:
account_leaf folder

miha-stopar committed:
assignment of nibbles_count

miha-stopar committed:
branch folder

miha-stopar committed:
storage_leaf folder

miha-stopar committed:
chips -> configs

miha-stopar committed:
WitnessRow in a separate file

miha-stopar committed:
extension node nibbles count constraint

miha-stopar committed:
documenting branch.rs

miha-stopar committed:
assign_branch_init moved into branch.rs

miha-stopar committed:
assing_branch_child into branch.rs

miha-stopar committed:
branch and extension node number of nibbles constraints fixed

miha-stopar committed:
Only two nil-nodes when placeholder branch constraint fixed

miha-stopar committed:
assign_row moved into witness_row

miha-stopar committed:
assing_branch_row moved into witness_row

miha-stopar committed:
Account address has 64 nibbles constraint

miha-stopar committed:
Documenting branch

miha-stopar committed:
BranchCols, ProofTypeCols structs moved

miha-stopar committed:
MainCols moved

miha-stopar committed:
AccumulatorPair and AccumulatorCols moved

miha-stopar committed:
DenoteCols moved

miha-stopar committed:
storage leaf key nibbles count constraint

miha-stopar committed:
Fixing comments

miha-stopar committed:
proof_type assignment moved to witness_row

miha-stopar committed:
assign_lookup_columns

miha-stopar committed:
nibbles_count = 64 for storage leaf all constraints implemented

miha-stopar committed:
fixing drifted_pos related constraints (placeholder branch)

miha-stopar committed:
branch is_drifted_pos/modified_node constraints fixed

miha-stopar committed:
branch modified_node/drifted_pos comments; cleaning

miha-stopar committed:
branch.rs comments finished

miha-stopar committed:
comments for branch_hash_in_parent

miha-stopar committed:
branch_init documented

miha-stopar committed:
is_codehash_proof reintroduced - but only read op, not modification

miha-stopar committed:
branch_key comments

miha-stopar committed:
Branch parallel docs

miha-stopar committed:
branch_rlc docs

miha-stopar committed:
extension node assign function

miha-stopar committed:
removed constraints for key_rlc & mult_diff being set in all branch children - needed only in branch init

miha-stopar committed:
documenting extension_node_key

miha-stopar committed:
extension node key docs

miha-stopar committed:
documenting extension node key

miha-stopar committed:
documenting extension node

miha-stopar committed:
extension_node comments and simplification

miha-stopar committed:
storage leaf key comments

miha-stopar committed:
documenting storage leaf key

miha-stopar committed:
storage leaf key constraints simplification

miha-stopar committed:
account leaf key simplification

miha-stopar committed:
storage leaf key assign function added

miha-stopar committed:
storage leaf value assign function

miha-stopar committed:
storage leaf key in added branch assign function

miha-stopar committed:
fixed bug in leaf/account key in added branch (introduced when clearing assignments and constraints for branch / extension node)

miha-stopar committed:
mostly renaming

miha-stopar committed:
Previous key RLC in branches removed (no PoisonedConstraints anymore)

miha-stopar committed:
Account leaf added in new branch simplified

miha-stopar committed:
storage leaf in added branch simplified

miha-stopar committed:
get_type used in account leaf assign functions

miha-stopar committed:
codehash modification enabled

miha-stopar committed:
Storage leaf key documented

miha-stopar committed:
documenting leaf_value

miha-stopar committed:
leaf_value.rs documented

miha-stopar committed:
Documenting and refactoring leaf_key_in_added_branch

miha-stopar committed:
leaf_key_in_added_branch documented; missing constraints added

miha-stopar committed:
Branch hash in first storage trie level lookup moved to branch_hash_in_parent.rs

miha-stopar committed:
Extension node in first storage trie level lookup moved to extension_node.rs

miha-stopar committed:
Only storage leaf hash is storage trie root

miha-stopar committed:
leaf value fix

miha-stopar committed:
Constraints for the only storage leaf which is also a placeholder moved into leaf_value.rs

miha-stopar committed:
storage_root_in_account_leaf.rs removed, all constraints moved to other files

miha-stopar committed:
counter removed: lookups will use start_root/end_root instead

miha-stopar committed:
not_first_level constraints

miha-stopar committed:
start_root, final_root constraints refactored

miha-stopar committed:
ensuring there is account proof before storage proof: documented and missing constraint added

miha-stopar committed:
Selectors documented and missing constraint added

miha-stopar committed:
formatting

miha-stopar committed:
formatting leaf_value.rs

miha-stopar committed:
PositionCols (q_enabled, q_not_first, not_first_level)

miha-stopar committed:
leaf_key_in_added_branch formatting

miha-stopar committed:
leaf_key clones removed

miha-stopar committed:
selectores redundant clones removed

miha-stopar committed:
proof_chain redundant clones

miha-stopar committed:
Pointers to constraints for the first level; ProofVariables renamed

miha-stopar committed:
resolving clippy warnings

miha-stopar committed:
resolving clippy warnings

miha-stopar committed:
resolving clippy warnings

miha-stopar committed:
resolved clippy warnings in extension_node.rs

miha-stopar committed:
resolved clippy warnings in branch configs

miha-stopar committed:
clippy warnings resolved in account configs

miha-stopar committed:
last clippy warnings resolved (mostly related to visibility)

miha-stopar committed:
long value RLP check fixed

miha-stopar committed:
rotation fixed (hash of the empty storage trie)

miha-stopar committed:
leaf key in added branch fix

miha-stopar committed:
tests regenerated

miha-stopar committed:
mpt moved into zkevm-circuits

miha-stopar committed:
println removed

miha-stopar committed:
power_of_randomness as instance

miha-stopar committed:
cleaning related to power_of_randomness

miha-stopar committed:
switching to kecccak_circuit (branch_hash_in_parent.rs ported)

miha-stopar committed:
account_leaf_storage_codehash.rs ported to keccak circuit

miha-stopar committed:
account_leaf_key_in_added_branch.rs ported to keccak circuit

miha-stopar committed:
extension_node ported to keccak circuit

miha-stopar committed:
storage leaf ported to keccak circuit

miha-stopar committed:
is_big_endian added to dev_load for keccak table construction

miha-stopar committed:
fix in keccak assignments

miha-stopar committed:
power_of_randomness moved

miha-stopar committed:
randomness now assign parameter

miha-stopar committed:
fix in get_branch_len

miha-stopar committed:
proof_type column added; constraints for proof_type properly correspond to boolean proof selectors

miha-stopar committed:
enabling all lookups in the same set of columns

miha-stopar committed:
lookup selector constraints (to prevent lookups in wrong rows)

miha-stopar committed:
enabling lookups

miha-stopar committed:
storage mod lookup enabled in leaf value row

miha-stopar committed:
mpt_table exposed

miha-stopar committed:
codehash lookup fix

miha-stopar committed:
clippy warnings resolved

miha-stopar committed:
key_len_lookup enabled; tests passed

miha-stopar committed:
check_zeros parameter to enable/disable 2^13 fixed table

miha-stopar committed:
row added to the storage leaf for non existing storage proof

miha-stopar committed:
non_existing_storage skeleton prepared

miha-stopar committed:
non-existing-storage proof now works for short (one RLP byte) leaves

miha-stopar committed:
storage-non-existing long (two RLP bytes) key enabled

miha-stopar committed:
key_len_lookups and number of nibbles enabled for long keys (non existing storage proof)

miha-stopar committed:
storage-non-existing long key fixed; test added

miha-stopar committed:
non-existing-storage fix; tests added

miha-stopar committed:
Non-existing-storage for proofs with nil in branch

miha-stopar committed:
order of rows constraints fixed (after non-existing-row for storage was added)

miha-stopar committed:
non-existing-storage docs

miha-stopar committed:
tests updated

miha-stopar committed:
cargo fmt applied

miha-stopar committed:
formatting

miha-stopar committed:
tests regenerated

miha-stopar committed:
clippy

miha-stopar committed:
cargo fmt

Brechtpd committed:
[MPT] Misc refactoring (#972)

Some refactoring which I believe decreases code duplication and
increases code readability.

Some TODOs:
- [x] The code required to get the previous rlc/mult data or the
inclusion in the parent check is quite complicated because of the
different node types. Instead it may be better to use a fixed location
to store this data in so a node can simply use this data directly
instead of having to figure out on its own where to find the data. This
is cleaner because this way each node can decide on its own how these
should be handled.
- [x] Currently RLP decoding is done using a couple of selectors that
are inputs from the prover. Then there are some checks if these are
correct, though these are not complete. I think it's easier to think
about this if we would just use a lookup to directly verify if these
selectors are set correctly so we don't have to worry about edge cases
are cases that are hard to constrain using custom gates.
- [x] There are currently many cases in the main state maching because
each row type is it's own state. However there is not really any reuse
between custom gates between these rows, except for branches. It'll
likely be quite a bit simple to just have a single state for account,
storage and extension and just use multiple rows in those states
- [x] May be a good idea to split up branches and extension nodes. (semi
done)
- [ ] There's a couple of circuit tools that were added in this PR to
make writing the MPT circuit more manageable. A lot of these tools can
still be greatly improved.
- [ ] Number of lookups has been reduced a lot, but there are still many
optimization possibilities (have not looked into reducing the expression
degree for example).
- [x] The circuit uses a fixed layout which uses around 100 columns (of
which a lot need to be byte constrained so needs a lot of lookups as
well). This makes the circuit quite a bit more dense than probably
required, a more flexible way to manager the required data so the
width/height can be choses would be very useful I think.

miha-stopar commented on zkevm-circuits/src/mpt_circuit/storage_leaf.rs:
```diff
@@ -0,0 +1,436 @@
+            rlp_values[StorageRowType::KeyS as usize].clone(),
+            rlp_values[StorageRowType::KeyC as usize].clone(),
+        ];
+        let value_item = [
```
@Brecht: It seems now the RLP bytes length for storage value is always 1. This actually never happens - because it's either 0 (when value is of length 1) or 2 (when value is of length bigger than 1). None of the constraints fail because it affects only the RLC of the value, which gets stored into the MPT table and currently we have no lookups into it yet.

Shall I modify this?

Brechtpd replied:
I had the storage values wrong for a long while so it wouldn't surprise me if I still have it wrong! 

Seems like I at least documented what I think should happen above:

>  // Because the storage value is an rlp encoded string inside another rlp encoded
// string (leaves are always encoded as [key, value], with
// `value` here containing a single stored value) the stored
// value is either stored directly in the RLP encoded string if short, or stored
// wrapped inside another RLP encoded string if long.

So the RLP value stored here is the internal value, not the main RLP value (see line 114 for that one). So this one will only be used when the RLP value is long. I hope this makes sense.

miha-stopar replied:
Ah, true! The value was correctly computed!

However, I find a bit confusing that in the case of the short RLP, the descriptor contains something like `value_rlp_bytes":[[1],[161]]`. In this example, the previous value RLP is short and the current value RLP is long. Thus, in the short case, `value_rlp_bytes` stores the actual value (being `1` in this case) and not the RLP byte, while the `value` is `[0,0,...]`.

In https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1353/commits/92ef471f51e5c70b7e3ececa7a2a3b3b34d172fb I changed to `value_rlp_bytes":[[],[161]]`
and `value` to be `[1,0,...]`. Can you have a look if it seems right to you?

Brechtpd replied:
I think this could work, but I also think it complicates things a bit as well because it breaks the meaning of what "short" is. `value_rlp_bytes":[[1],[161]]` for me makes sense (well, RLP encoding itself is a bit crazy) because the `1` is indeed the RLP byte, it also just happens to be the value.

So moving the `1` to the value bytes normally used for long values I think works, but also a bit non-standard compared to how short values are handled elsewhere (`RLPValueGadget` functions currently would return the wrong values for most things, though I guess this would be fixable by having a special mode for no data, instead of reusing the short flag).

Maybe another argument is that the storage leaf code works similarly to how the account leaf code works before, it's just that the account leaf always stores multiple values so the values stored in `value_rlp_bytes` are always actual RLP bytes.

But I don't really have a strong opinion on it.

miha-stopar replied:
My impression was that `value_rlp_bytes` is used for RLP bytes (the ones that specify the length, not the actual value). But I don't have a strong opinion either and thus I reverted to the previous (your) version. Thanks, Brecht!

Brechtpd committed:
Sync MPT to main (#1444)

### Description

This PR simply merges the current `main` into the MPT branch `mpt2`.
This is the first step to get some further MPT changes merged into the
MPT branch without making the PRs too big.

```
This PR contains:
- Merge `main` into `mpt2` with the necessary changes to the MPT circuit
```

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

Intermediate step, so limited testing

---------

Signed-off-by: smtmfft <smtm@taiko.xzy>
Signed-off-by: smtmfft <smtm@taiko.xyz>
Co-authored-by: pinkiebell <40266861+pinkiebell@users.noreply.github.com>
Co-authored-by: Haichen Shen <shenhaichen@gmail.com>
Co-authored-by: z2trillion <z2trillion@users.noreply.github.com>
Co-authored-by: Eduard S <eduardsanou@posteo.net>
Co-authored-by: adria0.eth <5526331+adria0@users.noreply.github.com>
Co-authored-by: Han <tinghan0110@gmail.com>
Co-authored-by: smtmfft <99081233+smtmfft@users.noreply.github.com>
Co-authored-by: smtmfft <smtm@taiko.xzy>
Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>
Co-authored-by: DreamWuGit <wwuwwei@126.com>
Co-authored-by: Rohit Narurkar <rohit.narurkar@protonmail.com>
Co-authored-by: Zhang Zhuo <mycinbrin@gmail.com>
Co-authored-by: Ming <hero78119@gmail.com>
Co-authored-by: Wu Sung Ming <wusm@sea.com>
Co-authored-by: xiaodino <qinyi@taikocha.in>
Co-authored-by: Mason Liang <mason@0x.org>
Co-authored-by: Steven <asongala@163.com>
Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>
Co-authored-by: Paul <108982045+ChengYueJia@users.noreply.github.com>
Co-authored-by: AronisAt79 <45406504+AronisAt79@users.noreply.github.com>
Co-authored-by: Mason Liang <mason@scroll.io>
Co-authored-by: JohnWick2ETH <104916412+JohnWick2ETH@users.noreply.github.com>
Co-authored-by: kunxian xia <xiakunxian130@gmail.com>
Co-authored-by: Leo Lara <leolara@users.noreply.github.com>
Co-authored-by: zengzengzenghuy <85567868+zengzengzenghuy@users.noreply.github.com>
Co-authored-by: john xu <john@taiko.xyz>
Co-authored-by: Miha Stopar <miha.stopar@xlab.si>
Co-authored-by: aguzmant103 <67167307+aguzmant103@users.noreply.github.com>
Co-authored-by: Richord <rzhang139@gmail.com>
Co-authored-by: Kimi Wu <leonhartwu@gmail.com>
Co-authored-by: Mickey <hedgefund996@gmail.com>
Co-authored-by: Luozhu <70309026+LuozhuZhang@users.noreply.github.com>
Co-authored-by: DoHoon Kim <59155248+DoHoonKim8@users.noreply.github.com>
Co-authored-by: jeong0982 <jeong0982@kaist.ac.kr>
Co-authored-by: Andy Arditi <andyrdt@gmail.com>
Co-authored-by: David Nevado <davidnevadoc@users.noreply.github.com>
Co-authored-by: omahs <73983677+omahs@users.noreply.github.com>
Co-authored-by: Richord <rrzhang@usc.edu>
Co-authored-by: ntampakas <nick@gmx.co.uk>
Co-authored-by: ashWhiteHat <phantomofrotten@gmail.com>
Co-authored-by: naure <naure@users.noreply.github.com>
Co-authored-by: Aurélien Nicolas <info@nau.re>
Co-authored-by: testuser-at-490752553772 <m_aronis@hotmail.com>
Co-authored-by: xiaodino <ruby@taiko.xyz>
Co-authored-by: Thomas Pan <43927782+thomaspanf@users.noreply.github.com>
Co-authored-by: Akase Cho <lightsing@users.noreply.github.com>
Co-authored-by: Vu <vu.voth@gmail.com>
Co-authored-by: Alex Beregszaszi <alex@rtfs.hu>
Co-authored-by: Enrico Bottazzi <85900164+enricobottazzi@users.noreply.github.com>
Co-authored-by: adria0 <adria0@nowhere>

Brechtpd committed:
Merge branch 'main' into mpt2

# Conflicts:
#	Cargo.lock
#	circuit-benchmarks/Cargo.toml
#	eth-types/src/geth_types.rs
#	gadgets/src/util.rs
#	keccak256/Cargo.toml
#	keccak256/src/arith_helpers.rs
#	keccak256/src/keccak_arith.rs
#	keccak256/src/lib.rs
#	keccak256/src/plain.rs
#	testool/src/compiler.rs
#	zkevm-circuits/Cargo.toml
#	zkevm-circuits/src/evm_circuit.rs
#	zkevm-circuits/src/pi_circuit.rs
#	zkevm-circuits/src/state_circuit.rs
#	zkevm-circuits/src/state_circuit/constraint_builder.rs
#	zkevm-circuits/src/table.rs
#	zkevm-circuits/src/table/keccak_table.rs
#	zkevm-circuits/src/table/mpt_table.rs
#	zkevm-circuits/src/tx_circuit/dev.rs
#	zkevm-circuits/src/tx_circuit/sign_verify.rs
#	zkevm-circuits/src/util.rs
#	zkevm-circuits/src/witness/mpt.rs

Brechtpd committed:
Merge branch 'main' into mpt2

Brechtpd committed:
MPT improvements (#1462)

### Description

- Added some missing MPT constraints
- Integrated some EVM tools like degree reduction and lookups using
dedicated columns

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

---------

Co-authored-by: Cecilia Zhang <zhangyixin319@gmail.com>
Co-authored-by: CeciliaZ030 <45245961+CeciliaZ030@users.noreply.github.com>

miha-stopar commented on zkevm-circuits/src/mpt_circuit/account_leaf.rs:
```diff
@@ -0,0 +1,589 @@
+                // - 2. Account leaf is deleted from a branch with two leaves, the remaining
+                // leaf moves one level up and replaces the branch. In this case we
+                // have a branch placeholder.
+                // TODO(Brecht): For case 2: just having the parent branch be the placeholder seems not enough
```
@Brechtpd The other constraints are covered by `DriftedGadget`, right?

Brechtpd replied:
Yeah seems ok to me now, not sure if I thought of any other problems

miha-stopar commented on zkevm-circuits/src/mpt_circuit/helpers.rs:
```diff
@@ -0,0 +1,1382 @@
+#[derive(Clone, Debug, Default)]
+pub struct IsEmptyTreeGadget<F> {
+    is_in_empty_trie: IsEqualGadget<F>,
+    is_in_empty_branch: IsEqualGadget<F>,
```
@Brechtpd Not that important, but perhaps we could use some other name instead of `is_in_empty_trie` and `is_in_empty_branch`? Seems a bit strange to be something in the empty something.

Brechtpd replied:
Any suggestions? For me it kind of makes sense, but I can understand how it could be confusing, though I can't immediately come up with a different name.

miha-stopar replied:
It seems to me it would be easier to read if we would have:
```
pub struct IsPlaceholderLeafGadget<F> {
    is_empty_trie: IsEqualGadget<F>,
    is_nil_in_branch_at_mod_index: IsEqualGadget<F>,
}
```
instead of
```
pub struct IsEmptyTreeGadget<F> {
    is_in_empty_trie: IsEqualGadget<F>,
    is_in_empty_branch: IsEqualGadget<F>,
}
```

And `is_placeholder_leaf` instead of `is_in_empty_trie` in `account_leaf.rs` and `storage_leaf.rs`.

What do you think?

miha-stopar commented on zkevm-circuits/src/mpt_circuit/helpers.rs:
```diff
@@ -0,0 +1,1382 @@
+        is_non_existing: Expression<F>,
+        key_value: &RLPItemView<F>,
+        key_rlc: &Expression<F>,
+        wrong_item: &RLPItemView<F>,
```
@Brechtpd Maybe `wrong_item` is a bit misleading as it actually contains the address of the enquired address? The "wrong" address is actually stored in the `key_rlc`.

Brechtpd commented on zkevm-circuits/src/mpt_circuit/account_leaf.rs:
```diff
@@ -0,0 +1,589 @@
+
+            // Anything following this node is below the account
+            // TODO(Brecht): For non-existing accounts it should be impossible to prove
+            // storage leaves unless it's also a non-existing proof?
```
It's been quite a while since I wrote this so I'm trying to reconstruct why I thought that so I may have been wrong then and I most definitely could be wrong now. :)

- Good point on the address not matching!
- For placeholder leafs the account data passed in by the prover could be anything I think, so the storage root hash could be any value passed into the circuit by the prover no?

Do we need to support `StorageDoesNotExist` for non-existing accounts?

miha-stopar replied:
When there is a `NonExistingAccount` proof, any `StorageChanged` lookup into `MPTTable` would fail (as `proof_type` is part of the table).

If the attacker would try to prepare a `StorageChanged` proof with a non-existing account, the address wouldn't match when there is a wrong leaf. But indeed I am not sure whether the proof would fail in the case of the placeholder leaf. I think we would need to add a constraint to allow the placeholder leaf only for the `NonExistingAccount` proof. I can add this in the PR I am preparing.

Regarding `StorageDoesNotExist` for non-existing accounts - I think we don't need it, but @ed255 would know better.

miha-stopar committed:
MPT new descriptors & benchmark (#1492)

### Description

New descriptors are used which have `Vec<Node>` structure and thus the
conversion in `witness_row.rs` from `[][]byte` to `Vec<Node>` was
removed. Note that `keccak_data` is now part of the `Node` struct - each
node holds the byte streams that need to be hashed.

Additionally, an MPT benchmark was added (some structs needed to be made
public).

### Type of change

- [x] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [x] This change requires a documentation update

### Contents

- All tests updated.
- `prepare_witness` function removed from `witness_row.rs`.
- MPT test updated to not use `prepare_witness`.
- Benchmark added.

miha-stopar commented on zkevm-circuits/src/table/mpt_table.rs:
```diff
@@ -1,68 +1,134 @@
+    BalanceChanged = AccountFieldTag::Balance as isize,
     /// Code hash exists
-    CodeHashMod = AccountFieldTag::CodeHash as isize,
+    CodeHashExists = AccountFieldTag::CodeHash as isize,
```
@Brechtpd At some point I thought we will need only `CodeHashExists`, but after some discussions we realised we also need `CodeHashChanged`. Is it ok to change the name?

Brechtpd replied:
Oh yes of course, will make it part of the next batch of changes together with the name changes above!

miha-stopar commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -0,0 +1,719 @@
+                    }};
+                    // Main state machine
+                    matchx! {
+                        a!(state_machine.is_start) => {
```
@Brechtpd I think there are no constraints currently about which nodes can follow which nodes, but after some thought I think these are indeed not necessary as (in `account_leaf.rs`)
```
require!(config.main_data.is_below_account => false);
```
guarantees there are no two or more account leaves. Another possible attack would be to have two storage proofs one after another, but the probability is again negligible as the second storage trie would need to hash to 0 as (the first) `storage_leaf` stores the default values (`parent_data.rlc = 0`). So I think it's all good :)

Brechtpd replied:
Yes I also believe it should be fine but could miss some case. I actually thought `is_below_account` was set to false again after the storage leaf, but looks like that isn't the case so seem like the hash being `0` is indeed the only thing stopping it from happening. Perhaps it can't hurt to also update `is_below_account` though. :)

miha-stopar commented on zkevm-circuits/src/mpt_circuit/rlp_gadgets.rs:
```diff
@@ -0,0 +1,770 @@
+    pub(crate) bytes: Vec<u8>,
+}
+
+impl<F: Field> RLPListGadget<F> {
```
If I am not mistaken, there are no constraints that would ensure that the byte streams are properly RLP encoded currently. So, for example for a storage leaf
`
[226 160 59 138 106 70 105 186 37 13 38 205 122 69 158 202 157 33 95 131 7 227 58 235 229 3 121 188 90 54 23 236 52 68 1]
`
the RLP gadget decodes `226` to `is_list=true` and `is_short=true`, from which it then knows how to compute for example the number of the RLP bytes. But the byte stream like
`
[227 160 59 138 106 70 105 186 37 13 38 205 122 69 158 202 157 33 95 131 7 227 58 235 229 3 121 188 90 54 23 236 52 68 1 x]
`
would be accepted as well (as far as RLP gadget is concerned), although you would still need to find the hash collision to make the attack work.

In the old code I had constraints like
`
226 - 192 = 160 - 128 + 2
`
But indeed because of the hashing these might not be necessary.

@Brechtpd  Do you think that checking that the byte streams are properly RLP encoded is not necessary?

miha-stopar replied:
Ah, yes, they are there! Sorry, my brain was formatted too much during vacation :)

Brechtpd committed:
MPT: merge with main only (#1530)

### Description

Just merges the MPT branch with the master, without fixing any issues
(and so breaks the code). The actual merge is done in
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1531,
the only purpose of this PR is to have a nicer diff for the other PR
without breaking the code in the current `mpt2` branch.

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

Brechtpd committed:
Merge branch 'main' into mpt2

# Conflicts:
#	.github/workflows/test-features.yml
#	Cargo.lock
#	bus-mapping/src/circuit_input_builder.rs
#	bus-mapping/src/evm/opcodes.rs
#	bus-mapping/src/evm/opcodes/create.rs
#	integration-tests/Cargo.toml
#	integration-tests/src/bin/gen_blockchain_data.rs
#	testool/Cargo.toml
#	zkevm-circuits/src/bytecode_circuit/circuit.rs
#	zkevm-circuits/src/bytecode_circuit/test.rs
#	zkevm-circuits/src/copy_circuit/dev.rs
#	zkevm-circuits/src/evm_circuit.rs
#	zkevm-circuits/src/evm_circuit/execution.rs
#	zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs
#	zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs
#	zkevm-circuits/src/evm_circuit/execution/push.rs
#	zkevm-circuits/src/evm_circuit/execution/stop.rs
#	zkevm-circuits/src/keccak_circuit.rs
#	zkevm-circuits/src/keccak_circuit/cell_manager.rs
#	zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs
#	zkevm-circuits/src/state_circuit.rs
#	zkevm-circuits/src/table.rs
#	zkevm-circuits/src/table/bytecode_table.rs
#	zkevm-circuits/src/table/mpt_table.rs
#	zkevm-circuits/src/util.rs
#	zkevm-circuits/src/util/cell_manager.rs
#	zkevm-circuits/src/witness/block.rs
#	zkevm-circuits/src/witness/bytecode.rs
#	zkevm-circuits/src/witness/mpt.rs

Brechtpd committed:
Removed file that was deleted on main

Brechtpd committed:
PR feedback

ChihChengLiang reviewed (commented):
I see the PR is getting closer to being merged.
@miha-stopar @Brechtpd @CeciliaZ030 Epic work on MPT!
Question for any of you: Is there anything you think the reviewers can focus on reviewing it? Otherwise, we can approve directly and unblock the merge.

Currently, I only quickly checked the files to see if any leftover files that shouldn't be included. The PR looks great for this part.

Brechtpd commented:
I think the most important thing is that the non-MPT specific changes are all good to go. If that's okay I think should be okay to merge in so it's easier to keep in sync/improve on/integrate better with the other code afterwards.

For a bit deeper review, I think also good to take a quick look at the main circuit implementation files (`mpt_circuit.rs`, `account_leaf.rs`, `storage_leaf.rs`, `extension_branch.rs`, `branch.rs`, `extension.rs`) to get a feeling how the circuit works.

I think the other files don't need much looking into for now unless to check how something works a bit more deeply. They are used only in the MPT circuit and the code there should be shared better with the other circuits once this PR is merged in (quite a bit of duplication there now to prevent merge conflicts and keep the diff clean).

Brechtpd committed:
Fix a comment

ChihChengLiang reviewed (approved):
LGTM!
@Brechtpd Thanks for the reviewing guide!
